Riot/Modules/Common/Recents/DataSources/RecentsDataSource.m

```diff
@@ -165,6 +150,59 @@ - (DiscussionsCount *)groupMissedDiscussionsCount
     return self.recentsListService.conversationMissedDiscussionsCount;
 }
 
+#pragma mark - Sections
+
+- (RecentsDataSourceSections *)makeDataSourceSections
```

Comment:
This is the core logic of constructing sessions that is moved from `numberOfSectionsInTableView`. Instead of setting the indexes directly it will just construct an array of types before creating the sections.

I also extracted this code into its separate method so that it can be called at any point.